src/main/webapp/teacher/controlGameInstance/controlGameInstance.js

```diff
@@ -235,4 +361,87 @@ function initUIControlButtons(gameInstanceId) {
   });
 }
 
+// Probably should do this following the Linked List instead of querySnapshot.forEach
+// Get all questions in the "questions" collection in the gameInstance
+function buildQuestionHistory(gameInstanceId) {
+  const questionInGameInstanceCollectionRef = db.collection('gameInstance').doc(gameInstanceId).collection('questions');
+  questionInGameInstanceCollectionRef.get().then(function(querySnapshot) {
+    querySnapshot.forEach(function(doc) {
+      // doc.data() is never undefined for query doc snapshots
+      const questionId = doc.id
+
+      // Add the question to the UI
+      addQuestionToHistoryUI({ question: doc.data(), questionId });
+
+      // Add the question's answers to the UI
+      buildQuestionAnswersHistory({ questionId, questionsCollectionRef: questionInGameInstanceCollectionRef });
+    });
+  });
+}
+
+// Add a question to the Question History section of the UI 
+function addQuestionToHistoryUI({ question, questionId } = {}) {
+  const questionStatsDivElement = document.getElementById('jsQuestionStats');
+  
+  // The div for this new question element
+  const singleQuestionStatDivElement = document.createElement('div');
+  singleQuestionStatDivElement.id = 'stats-' + questionId
+  singleQuestionStatDivElement.classList.add('questionInHistory');
+
+  const questionTitle = document.createElement('div');
+  questionTitle.innerText = 'Question title: ' + question.title;
+  singleQuestionStatDivElement.appendChild(questionTitle);
+
+  // How many students have answered the question
+  const numberQuestionAnswersHistoryElement = document.createElement('div');
+  numberQuestionAnswersHistoryElement.id = 'stats-' + questionId + '-numberQuestionAnswers';
```

Comment:
What are these ids used for? Later retrieval?

Reply:
Yep, in line 320 I update this DOM element and access it with its ID. All of this is so that I only build the question history "tree" once, and when that question becomes active and is updated as students answer, I just go and update that number in the history section as well, instead of having to recreate the whole question history object. Do you think it's OK or too complicated?